flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayComparableElementTypeStrategy.java

```diff
@@ -0,0 +1,147 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.ArgumentCount;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.inference.ConstantArgumentCount;
+import org.apache.flink.table.types.inference.InputTypeStrategy;
+import org.apache.flink.table.types.inference.Signature;
+import org.apache.flink.table.types.logical.LegacyTypeInformationType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.StructuredType.StructuredComparison;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * An {@link InputTypeStrategy} that checks if the input argument is an ARRAY type and check whether
+ * its' elements are comparable.
+ *
+ * <p>It requires one argument.
+ *
+ * <p>For the rules which types are comparable with which types see {@link
+ * #areComparable(LogicalType, LogicalType)}.
+ */
+@Internal
+public final class ArrayComparableElementTypeStrategy implements InputTypeStrategy {
+    private final StructuredComparison requiredComparison;
+    private final ConstantArgumentCount argumentCount;
+
+    public ArrayComparableElementTypeStrategy(StructuredComparison requiredComparison) {
+        Preconditions.checkArgument(requiredComparison != StructuredComparison.NONE);
+        this.requiredComparison = requiredComparison;
+        this.argumentCount = ConstantArgumentCount.of(1);
+    }
+
+    @Override
+    public ArgumentCount getArgumentCount() {
+        return argumentCount;
+    }
+
+    @Override
+    public Optional<List<DataType>> inferInputTypes(
+            CallContext callContext, boolean throwOnFailure) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+        List<LogicalType> argumentTypes =
+                argumentDataTypes.stream()
+                        .map(DataType::getLogicalType)
+                        .collect(Collectors.toList());
+
+        if (!argumentTypes.stream()
+                .allMatch(logicalType -> logicalType.is(LogicalTypeRoot.ARRAY))) {
+            return callContext.fail(throwOnFailure, "All arguments requires to be a ARRAY type");
+        }
```

Comment:
Because I fixed the argument, I set it one, so it the input is not one argument, it will throw exception such as
`No match found for function signature ARRAY_MAX(<INTEGER ARRAY>, <INTEGER ARRAY>)`

Reply:
What I meant is: 
Let's take `ARRAY_MAX(<INTEGER ARRAY>, <INTEGER ARRAY>, <INTEGER ARRAY>)`. Given the current code you perform two unnecessary checks. It does not matter if second and third arguments are arrays or not.

If you do:
```
if (argumentDataTypes.size() == 1) {
            final DataType argumentType = argumentDataTypes.get(0);
            if (!argumentType.getLogicalType().is(LogicalTypeRoot.ARRAY)) {
                return callContext.fail(throwOnFailure, "All arguments requires to be an ARRAY type");
            }
            final DataType elementDataType =
                    ((CollectionDataType) argumentType).getElementDataType();
} else {
  return callContext.fail(...)
}
```
you omit a few potential checks.

To be honest we don't even need to check the size of arguments, because that is already done by the `ArgumentCount`. We could even simplify the entire code to:

```
    public Optional<List<DataType>> inferInputTypes(
            CallContext callContext, boolean throwOnFailure) {
        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
        final DataType argumentType = argumentDataTypes.get(0);
        if (!argumentType.getLogicalType().is(LogicalTypeRoot.ARRAY)) {
            return callContext.fail(throwOnFailure, "All arguments requires to be an ARRAY type");
        }
        final DataType elementDataType =
                ((CollectionDataType) argumentType).getElementDataType();
        final LogicalType elementLogicalDataType = elementDataType.getLogicalType();
        if (!areComparable(elementLogicalDataType, elementLogicalDataType)) {
            return callContext.fail(
                    throwOnFailure,
                    "Type '%s' should support %s comparison with itself.",
                    elementLogicalDataType,
                    comparisonToString());
        }
        return Optional.of(argumentDataTypes);
    }
```

Reply:
Sorry, I don't understand your last message. Could you rephrase it?

Reply:
> you omit a few potential checks.

you said I omit a few potential checks. I don't know what these potential checks are.

Reply:
> public Optional<List<DataType>> inferInputTypes(
>             CallContext callContext, boolean throwOnFailure) {
>         final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
>         List<LogicalType> argumentTypes =
>                 argumentDataTypes.stream()
>                         .map(DataType::getLogicalType)
>                         .collect(Collectors.toList());
> 
>         if (!argumentTypes.stream()
>                 .allMatch(logicalType -> logicalType.is(LogicalTypeRoot.ARRAY))) {
>             return callContext.fail(throwOnFailure, "All arguments requires to be a ARRAY type");
>         }
>         if (argumentDataTypes.size() == 1) {
>             final DataType elementDataType =
>                     ((CollectionDataType) argumentDataTypes.get(0)).getElementDataType();
>             final LogicalType elementLogicalDataType = elementDataType.getLogicalType();
>             if (!areComparable(elementLogicalDataType, elementLogicalDataType)) {
>                 return callContext.fail(
>                         throwOnFailure,
>                         "Type '%s' should support %s comparison with itself.",
>                         elementLogicalDataType,
>                         comparisonToString());
>             }
>         }
>         return Optional.of(argumentDataTypes);
>     }
I can simplify this code block.

Reply:
> you said I omit a few potential checks. 

What I meant is that:
**IF** you check the `size` **FIRST** you skip checking potentially arbitrary big number of checks if an input type is an **ARRAY**

We do not want to perform unnecessary checks.

Reply:
Ok, got it. Thank you for your explain.